tools/src/main/java/org/wildfly/swarm/tools/DeclaredDependencies.java

```diff
@@ -246,13 +247,23 @@ public void writeTo(File file) {
     }
 
     @Override
-    protected int comparator(ArtifactSpec first, ArtifactSpec second) {
-        if (first.scope.equals("compile") || first.scope.equals("provided")) {
-            return -1;
-        } else if (second.scope.equals("compile") || second.scope.equals("provided")) {
-            return 1;
-        }
-        return 0;
+    public Collection<ArtifactSpec> getDirectDeps() {
+        // Reorder so that compile and provided come first, but otherwise keep the original order. This is important in
+        // case when the final collection is going to be resolved. Different order of direct dependencies may lead to
+        // different resolved transitive dependencies.
+
+        Collection<ArtifactSpec> originalDeps = super.getDirectDeps();
+        List<ArtifactSpec> reordered = new ArrayList<>(originalDeps.size());
+        originalDeps.stream().filter(s -> PRIORITIZED_SCOPES.contains(s.scope)).forEach(reordered::add);
+        originalDeps.stream().filter(s -> !PRIORITIZED_SCOPES.contains(s.scope)).forEach(reordered::add);
+        return reordered;
+    }
+
+    private static final List<String> PRIORITIZED_SCOPES = new ArrayList<>();
```

Comment:
Should be a `Set`?

Reply:
(I know it's tiny, but conceptually, it should be a `Set` :-) )

Reply:
Or really just an inline `Predicate`.

Reply:
Yeah, good point, I will make it a predicate :).

Reply:
I made it a Set.